test/suites/edgex-no-sec/ekuiper_test.go

```diff
@@ -64,4 +74,98 @@ func TestRulesEngine(t *testing.T) {
 
 		require.Greaterf(t, reading.TotalCount, 0, "No readings have been re-published to EdgeX message bus by ekuiper")
 	})
+
+}
+
+func subtestSetup(t *testing.T) (teardown func(), err error) {
+	start := time.Now()
+
+	teardown = func() {
+		log.Println("[SUBTEST TEARDOWN]")
+		utils.SnapDumpLogs(t, start, deviceVirtualSnap)
+		utils.SnapDumpLogs(t, start, ekuiperSnap)
+		utils.SnapDumpLogs(t, start, ascSnap)
+
+		log.Println("Removing installed snap:", !utils.SkipTeardownRemoval)
+		if !utils.SkipTeardownRemoval {
+			utils.SnapRemove(t, deviceVirtualSnap)
+			utils.SnapRemove(t, ekuiperSnap)
+			utils.SnapRemove(t, ascSnap)
+		}
+	}
+	defer teardown()
```

Comment:
If you want to do it this way, the caller suppose to call this as deferred function call, otherwise, it is called as soon as this function returns.

But I'd suggest to simply put the logic in t.Cleanup and remove the returned function. This makes the rest of the code simpler too because you won't need to call teardown on errors but rather raise errors which automatically triggers the cleanup.